### PR TITLE
[Core] Added event loop metrics for posts.

### LINKED ISSUE #14545: [metrics] Add ASIO event loop statistics collection
### Describe your feature request

Collect count, queueing time, and execution time statistics for asio handler posts, and periodically dump these statistics to the debugging state file.

### PR DESCRIPTION
**Pull request by clarkzinzow**
## Why are these changes needed?

We need better observability on the ASIO event loop, especially around how many handlers are currently in the queue, as well as queueing and execution time statistics. This PR wraps the boost asio `io_context` object with a proxy that collects such statistics, and adds a human-readable name to every event loop `.post()` call.

Most of the changes consist of converting `boost::asio::io_context` to `io_context_proxy` throughout Ray, and adding a human-readable name to each `.post()` call. The core changes are concentrated in `ray/common/asio/io_context.*`.

## Stats collected

These stats are collected globally:

1. Cumulative post count.
2. Current post count.
3. Mean queue time.
4. Max queue time.
5. Min queue time.
6. Cumulative queue time.
7. Mean execution time.
8. Cumulative execution time.

These stats are collected per handler:

1. Cumulative post count.
2. Current post count.
3. Mean execution time.
4. Cumulative execution time.

## Example stats output
The per-handler stats are sorted by cumulative post count.
```
Global stats: 542231 total (416 active)
Queueing time: mean = 22.348 ms, max = 58.165 ms, min = 1.160 us, total = 12118.019 s
Execution time:  mean = 29.658 us, max = 16.807 ms, min = 127.000 ns, total = 16.082 s
Handler stats:
        ObjectStoreNotification.StoreAdd: 200541 total (0 active)
                Queueing time: mean = 25.320 ms, max = 58.126 ms, min = 1.800 us, total = 5077.716 s
                Execution time: mean = 10.793 us, max = 4.471 ms, min = 127.000 ns, total = 2.164 s
        ObjectStoreNotification.StoreRemove: 109142 total (200 active)
                Queueing time: mean = 20.306 ms, max = 58.165 ms, min = 21.841 us, total = 2216.244 s
                Execution time: mean = 12.121 us, max = 2.714 ms, min = 1.036 us, total = 1.323 s
        CoreWorkerService.AddObjectLocationOwner: 66838 total (0 active)
                Queueing time: mean = 24.235 ms, max = 57.823 ms, min = 1.561 us, total = 1619.839 s
                Execution time: mean = 6.607 us, max = 1.062 ms, min = 857.000 ns, total = 441.566 ms
        NodeManagerService.PinObjectIDs: 54763 total (99 active)
                Queueing time: mean = 19.090 ms, max = 54.516 ms, min = 1.632 us, total = 1045.449 s
                Execution time: mean = 187.650 us, max = 4.264 ms, min = 49.534 us, total = 10.276 s
        CoreWorkerService.WaitForObjectEviction: 54662 total (102 active)
                Queueing time: mean = 19.142 ms, max = 54.769 ms, min = 1.505 us, total = 1046.337 s
                Execution time: mean = 27.392 us, max = 1.298 ms, min = 6.282 us, total = 1.497 s
        CoreWorkerService.RemoveObjectLocationOwner: 54462 total (95 active)
                Queueing time: mean = 19.827 ms, max = 54.987 ms, min = 1.160 us, total = 1079.808 s
                Execution time: mean = 5.059 us, max = 1.257 ms, min = 851.000 ns, total = 275.529 ms
        HeartbeatInfoGcsService.ReportHeartbeat: 301 total (0 active)
                Queueing time: mean = 13.007 ms, max = 53.258 ms, min = 2.312 us, total = 3.915 s
                Execution time: mean = 12.050 us, max = 169.034 us, min = 3.627 us, total = 3.627 ms
```

## Implementation notes

- **Inheritance over Composition:** By having the proxy derive from `boost::asio::io_context`, the proxy implementation is primarily limited to the `post()` function, and the application code changes were limited to (1) `boost::asio::io_context` --> `io_context_proxy` type conversion, (2) adding a handler name to post calls. This also allowed us to easily roll up statistics for event loops that were passed down an object hierarchy (e.g. the raylet main event loop being used by both the node manager and the object manager). If we had instead implemented the proxy as a wrapper with a `boost::asio::io_context` member, we would have had to implement a complete passthrough interface on the wrapper (not trivial) and/or add more application code changes, with the latter compounding by being required at each level of the object hierarchy.
- **Locks over Atomics:** Each stats struct is guarded by its own mutex in a fine-grained locking effort. I originally implemented this stats struct synchronization using atomics (min/max done via CAS loops), but the memory ordering required across struct members in order to provide consistent statistics made it seem like per-member atomics wasn't giving us much over a simple read-write mutex on the entire struct.

## Related issue number

Closes #14545 

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/common/asio/io_context.cc
Comment (clarkzinzow) [inline src/ray/common/asio/io_context.cc:45]: This helper allows us to move a given handler into the handler wrapper lambda, avoiding a potentially expensive handler copy, while remaining in C++11 land. With C++14's generalized lambda capture, this helper isn't necessary:
```cpp
auto wrapped_handler = [this, handler = std::move(handler), stats, name, start]() {
  // ...
};
```

File: src/ray/common/asio/io_context.h
Comment (clarkzinzow) [inline src/ray/common/asio/io_context.h:83]: These aren't yet used outside of the `StatsString` implementation, but seemed like they would be useful to expose publicly in case some consumers want to poke at the handler stats outside of the full stats string, e.g. if some consumer detects heavy load, it might only be interested in a subset of the event loop stats.

File: src/ray/common/asio/io_context.cc
Comment (clarkzinzow) [inline src/ray/common/asio/io_context.cc:104]: Although `name` and `start` are cheap to copy, we might want to move `stats` into the lambda as well.

**Comment by rkooo567** [comment]
Other than the file that contains `io_context_proxy`, everything else is refactoring right?

**Comment by rkooo567** [comment]
Also, do you have the microbenchmark result?

**Comment by rkooo567** [comment]
Also, cc @iycheng

**Comment by clarkzinzow** [comment]
@rkooo567 Yep most of the changes are in `ray/common/asio/io_context.*`, just updated the PR description to indicate as much.
 
>Also, do you have the microbenchmark result?

I won't have microbenchmark results until #14540 is fixed, since the microbenchmarks currently crash on that test.

**Comment by rkooo567** [comment]
Why don't we run it with that test disabled? (I think we can at least see the performance impact of this PR in that way?)

**Comment by clarkzinzow** [comment]
@rkooo567 I'm thinking that Ray crashing on that test is indicative of a bad Ray regression, not just a bug with that test, so I'm not sure if we'll get good signal in those benchmarking results anyway.

**Comment by rkooo567** [comment]
Hmm but what we care about is how much this change will impact the general performance of ray right? I don't think the separate ray regression will impact the comparison? I am okay to wait until that issue is fixed, but I think we shouldn't merge this PR without understanding the performance impact (which means merging this PR could be delayed quite a bit).

**Comment by clarkzinzow** [comment]
@rkooo567 To be clear, I'm not advocating for merging this PR without benchmarks, I'm saying that the P0 bug that's causing that one test to break will probably break others, or will probably cause skewed performance. I'd assumed that bug would get fixed while this is getting reviewed. But if we're cool with disabling that test and taking the benchmarking results at face value (assuming others don't fail), then I'm cool with that too! 😁

File: src/ray/common/asio/io_context.cc
Comment (rkooo567) [inline src/ray/common/asio/io_context.cc]: Is Reader/Writer lock meaningful here? I can imagine most of the time we write stats and read-only happens on the main thread once in a while, so it is not necessary? But I might be missing something.

File: src/ray/common/asio/io_context.cc
Comment (rkooo567) [inline src/ray/common/asio/io_context.cc:49]: Btw, this should happen after the config is initialized at each component. Otherwise, the won't be reflected. https://github.com/ray-project/ray/blob/2da58bb0214644177c3f9a9c59d03cc5fa0d1c47/src/ray/core_worker/core_worker.cc#L139 (I am mentioning just to make sure you didn't miss it)

File: src/ray/common/asio/io_context.cc
Comment (rkooo567) [inline src/ray/common/asio/io_context.cc:203]: Why don't we make this function inlined?

**Review by rkooo567 — COMMENTED** [review]
I only have a couple of comments.

File: src/ray/common/asio/io_context.cc
Comment (clarkzinzow) [reply src/ray/common/asio/io_context.cc:203]: Hmm I thought that most compilers would inline this and the other function implicitly, but I don't see the harm in doing it explicitly. I'll add the specifier.

File: src/ray/common/asio/io_context.cc
Comment (clarkzinzow) [reply src/ray/common/asio/io_context.cc:49]: That's a great point, and also applies to the [raylet code path](https://github.com/ray-project/ray/blob/2da58bb0214644177c3f9a9c59d03cc5fa0d1c47/src/ray/raylet/main.cc#L120). And there it's even worse, since [we instantiate the `io_context` before we initialize the config](https://github.com/ray-project/ray/blob/2da58bb0214644177c3f9a9c59d03cc5fa0d1c47/src/ray/raylet/main.cc#L99-L100)... I think that we're going to have to check the config flag on every post, unfortunately.

File: src/ray/common/asio/io_context.cc
Comment (ericl) [inline src/ray/common/asio/io_context.cc:86]: Can you use absl::GetCurrentTimeNanos() instead of this?

File: src/ray/common/asio/io_context.cc
Comment (ericl) [inline src/ray/common/asio/io_context.cc:129]: How about we only store handler-specific stats for total execution time (drop queueing, and exec min/max)? I don't think queueing is meaningful on a per-handler basis. Min/max is probably overkill for our purposes.

File: src/ray/common/asio/io_context.cc
Comment (ericl) [inline src/ray/common/asio/io_context.cc:142]: Let's drop this, since it's captured by per-handler stats already.

File: src/ray/common/asio/io_context.cc
Comment (ericl) [inline src/ray/common/asio/io_context.cc:240]: Can we make this one line per item (remove queueing), and then add total to the this line inline? I think this will greatly improve readability at minimal info loss.

File: src/ray/common/asio/io_context.cc
Comment (ericl) [inline src/ray/common/asio/io_context.cc:52]: Feature flag :+1:

File: src/ray/common/asio/io_context.h
Comment (ericl) [inline src/ray/common/asio/io_context.h:53]: Can you rename this file to instrumented_io_context? And also the class accordingly? Proxy is less specific of a name.

File: src/ray/common/asio/io_context.cc
Comment (clarkzinzow) [reply src/ray/common/asio/io_context.cc:49]: Given that the config instance get, and the config flag get, should both be very fast (the latter is inlined), I think we'll be ok with checking the flag on every post.

If it actually was an expensive operation, ideally there would be an `is_initialized_` flag on `RayConfig` that's set after `RayConfig::instance()::initialize()` is called, along with an `is_initialized()` member function. Then, at the beginning of `post()` in `io_context_proxy`, we could do the following check:
```cpp
if (!config_initialized_) {
  config_initialized_ = RayConfig::instance().is_initialized();
  if (config_initialized_) {
    stats_collection_enabled_ = RayConfig::instance().asio_event_loop_stats_collection_enabled();
  }
}
if (!stats_collection_enabled_) {
  // Config is not initialized or stats collection is not enabled, either way, don't collect stats.
  return boost::asio::io_context::post(std::move(handler));
}
```
That way, once the config is initialized, each if statement should be easy to predict and result in hot dynamic branch prediction, and should be close to the same overhead as if we had set `stats_collection_enabled_` on instantiation, for the rest of `io_context_proxy`'s lifetime.

But again, I think `RayConfig::instance().asio_event_loop_stats_collection_enabled()` is cheap enough to call on each `post()`, so we should just do that. Do you agree?

File: src/ray/common/asio/io_context.cc
Comment (clarkzinzow) [reply src/ray/common/asio/io_context.cc]: My thinking was that there could be readers from multiple threads in the future (inspecting subsets of the stats concurrently), but that was probably premature. I'll change this to a pain lock for now.

File: src/ray/common/asio/io_context.cc
Comment (clarkzinzow) [reply src/ray/common/asio/io_context.cc:86]: Oh neat, I didn't know that we had standardized around that.

File: src/ray/common/asio/io_context.cc
Comment (clarkzinzow) [reply src/ray/common/asio/io_context.cc:240]: Would something like this work?
```
some_service.SomeMethod - 100 total (2 active), CPU time: mean = 500 us, total = 50 ms
```

File: src/ray/common/asio/io_context.cc
Comment (clarkzinzow) [reply src/ray/common/asio/io_context.cc:129]: Right now, execution time doesn't include queueing time. Should execution time instead be measured from the time at which the handler is enqueued?

File: src/ray/common/asio/io_context.cc
Comment (ericl) [reply src/ray/common/asio/io_context.cc:240]: Looks great.

File: src/ray/common/asio/io_context.cc
Comment (ericl) [reply src/ray/common/asio/io_context.cc:129]: The existing definition seems fine. Queueing is just not useful to measure at all at the handler level since it's an aggregate effect. You should get similar info from just global q stats.

Comment (clarkzinzow) [reply src/ray/common/asio/io_context.cc:129]: I agree, although I did find it useful as an indicator of "which handler is getting stuck in the queue for longer than it should".

**Comment by ericl** [comment]
Interesting. Wonder if we should sort by total cpu time, seems like a good
measure of heaviness.

**Comment by rkooo567** [comment]
So, it seems like the performance impact is pretty low even with the repetitive access to the global variable?

**Comment by rkooo567** [comment]
The current state LGTM if tests pass. But I will wait for other reviewers to approve

**Comment by ericl** [comment]
Lgtm

File: BUILD.bazel
Comment (rkooo567) [inline BUILD.bazel:366]: I guess we can remove this now? (if this was added for the Reader/Writer lock)

Comment (clarkzinzow) [reply BUILD.bazel:366]: Still needed for the `absl::Mutex` and the `absl::MutexLock`, I believe.

Comment (rkooo567) [reply BUILD.bazel:366]: Ah, I see. I thought it was already added somewhere since we used locks everywhere, but it looks like it was included in `ray_util`

Comment (clarkzinzow) [reply BUILD.bazel:366]: We're often relying on implicit linking via transitive dependencies for these kind of libraries, which makes a lot of these Bazel targets brittle when it comes to refactoring (e.g., if the `ray_util` dependency was removed, the `ray_common` library would break since it was implicitly relying on libraries included by `ray_util`). IMO, if a library (Bazel target) has a dependency like `absl/synchronization`, it should explicitly have it as a Bazel dependency.

**Comment by ericl** [comment]
Tests failing

**Review by ericl — APPROVED** [review]

**Review by fishbone — COMMENTED** [review]
I saw locks there and I have one concern about the overhead of calculating the metrics.

Is it better to use enum instead of string so we can pre-construct the stats array and use atomic counter there? 
Usually, lock is slower than atomic counter and atomic counter is 10x slower than raw int.
If we observe some issues with atomic we can try something else like thread-local count kind of thing.



For ad-hoc metrics, we can still use the lock in a current way.

File: src/ray/common/asio/io_service_pool.cc
Comment (iycheng) [inline src/ray/common/asio/io_service_pool.cc:28]: Just some thoughts here. Is it better to do things like:

`using ray_io_context = instrumented_io_context;`

and use `ray_io_context` everywhere? So later if we want to do this kind of change again, we don't need to update `instrumented_io_context` everywhere.

File: src/ray/common/asio/instrumented_io_context.h
Comment (clarkzinzow) [reply src/ray/common/asio/instrumented_io_context.h:73]: I initially started to go that route, since there are several benefits to that approach, being:

1. Collection (most likely an array) of stats structs could be allocated at instantiation time, and concurrent access could proceed without a collection-level mutex.
2. Enums would result in faster lookups (indexing into preallocated and immutable array) and better compile-time type checking than an opaque string key.

Unfortunately, the handler identifiers associated with the [client](https://github.com/ray-project/ray/blob/ca3ac543fe9717c574f40774fdecb58d69e15189/src/ray/rpc/grpc_client.h#L29-L33) and [server](https://github.com/ray-project/ray/blob/ca3ac543fe9717c574f40774fdecb58d69e15189/src/ray/rpc/grpc_server.h#L30-L37) gRPC methods are built dynamically via the macros, and defining an enum that contains all of those client and server gRPC methods would essentially spread the method definitions across very disparate modules.

The closest that I came to making that tenable was an enum definition at the beginning of this header file, with the macros passing the corresponding enum value to be used as the handler identifier. Do you think I should give that a go?

File: src/ray/common/asio/io_service_pool.cc
Comment (clarkzinzow) [reply src/ray/common/asio/io_service_pool.cc:28]: Also, I tend to grep `class Foo` when I'm trying to look up the definition of some class `Foo` that I'm seeing used somewhere, and I'm sure that `using` definition would throw me for a loop! 😁

File: src/ray/common/asio/instrumented_io_context.h
Comment (ericl) [reply src/ray/common/asio/instrumented_io_context.h:73]: A string is easier to deal with than an enum, and it appears the performance is good enough here. So I think we should keep it as a string.

Comment (clarkzinzow) [reply src/ray/common/asio/instrumented_io_context.h:73]: That's my opinion as well.

**Comment by clarkzinzow** [comment]
> Lint failing

@ericl Any particular reason buildifier isn't running with the rest of our formatting scripts, via `travis/ci/format.sh`?

File: src/ray/common/asio/io_service_pool.cc
Comment (iycheng) [reply src/ray/common/asio/io_service_pool.cc:28]: I agreed that `instrumented_io_context` does show more semantics about the purpose. The benefit of using is that later when we want to change it to something else, we only need to change that using statement. If we have that using at the beginning, we don't need so many files update of the types.


I'm OK with this. I guess later if we want to update it globally to something else, maybe we can abstract it. That's beyond the scope of this one.

Comment (clarkzinzow) [reply src/ray/common/asio/io_service_pool.cc:28]: >I agreed that instrumented_io_context does show more semantics about the purpose. The benefit of using is that later when we want to change it to something else, we only need to change that using statement. If we have that using at the beginning, we don't need so many files update of the types.

I totally get that reasoning, but since the users of `instrumented_io_context` are relying on an interface to use that instrumentation (`StatsString()` for an event loop stats summary, `post()` with an extra handler name argument), I think that having the abstraction's name convey the semantics of that interface is more of a good thing than a bad thing. If we change the semantics of the interface, then the name should be updated at that time to reflect those changed semantics, even if it has to be changed in a lot of places.

And thankfully code refactoring tools make these kind of name changes easy, even with basic shell tools:
```
find . -type f \( -name "*.h" -o -name "*.cc" \) -exec sed -i'' -e 's/instrumented_io_context/ray_io_context/g' {} +
```

File: src/ray/common/asio/instrumented_io_context.h
Comment (iycheng) [reply src/ray/common/asio/instrumented_io_context.h:73]: I guess it's because the bottleneck for the testing cases is not CPU for now. 
The same throughput doesn't mean they are using the same resources and given that we can have arbitrary functions, it can give different results under different workloads. The resource usage (mem/cpu) should also be included in the benchmark. Well, anyway, that's beyond the scope of this PR, so I'm fine with it as this now.

The string can introduce a typo kind of thing as well, maybe `const string&` in some central place to avoid literal string?

**Comment by clarkzinzow** [comment]
@ericl @rkooo567 I believe that the tests are good, although I haven't looked at the buildkite failure yet (will check tomorrow morning).

File: src/ray/common/asio/instrumented_io_context.h
Comment (clarkzinzow) [reply src/ray/common/asio/instrumented_io_context.h:73]: >The same throughput doesn't mean they are using the same resources, and given that we can have arbitrary functions, it can give different results under different workloads. The resource usage (mem/CPU) should also be included in the benchmark.

I strongly agree there. After we've fixed the current high priority core and IO bugs, I'm hoping that we can prioritize a detailed audit of certain hot paths for the raylet and core worker since I'm sure there's a lot of low-hanging fruit. I wouldn't be surprised if we identified this proxy as adding unnecessary overhead, at which point we can work towards minimizing that identified overhead. Also, if you'd like to help me out with implementing some core microbenchmarks testing the performance of core components (not in Python), or improving the existing Python microbenchmarks, I'd be happy to partner on that!

>The string can introduce a typo kind of thing as well, maybe const string& in some central place to avoid literal string?

I don't think that I follow, what would defining all string literals in a central place give us? Couldn't those literal string typos happen just as easily in that central place?

Comment (iycheng) [reply src/ray/common/asio/instrumented_io_context.h:73]: > I don't think that I follow, what would defining all string literals in a central place give us? Couldn't those literal string typos happen just as easily in that central place?

I originally thought that there might be a pattern like this:

```
// some place
post("metricsA")
// another place
post("metricsA")
```
With this pattern, it'll avoid typo kind of thing if we do it in this way

```
metricsA="metricsA"

// some place 
post(metricsA)
// another place
post(metricsA)
```
But it looks like it's not the case here. All posts are measuring different metrics.

Comment (clarkzinzow) [reply src/ray/common/asio/instrumented_io_context.h:73]: Ah got it, yeah each string literal should be defined once, at the point of posting.

**Pull request merged**
